src/bsd.rs

```diff
@@ -0,0 +1,126 @@
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::{
+    io::Error,
+    mem::{self, size_of},
+    net::IpAddr,
+    os::fd::{AsRawFd, FromRawFd, OwnedFd},
+    ptr, slice, str,
+};
+
+use libc::{
+    getpid, read, rt_msghdr, sockaddr_dl, sockaddr_in, sockaddr_in6, sockaddr_storage, socket,
+    write, AF_INET, AF_INET6, PF_ROUTE, RTAX_IFP, RTAX_MAX, RTA_DST, RTA_IFP, RTM_GET, RTM_VERSION,
+    SOCK_RAW,
+};
+
+use crate::{default_err, next_item_aligned_by_four};
+
+pub fn interface_and_mtu_impl(remote: IpAddr) -> Result<(String, usize), Error> {
+    // Open route socket.
+    let fd = unsafe { OwnedFd::from_raw_fd(socket(PF_ROUTE, SOCK_RAW, 0)) };
+    if fd.as_raw_fd() == -1 {
+        return Err(Error::last_os_error());
+    }
+
+    // Prepare buffer with destination `sockaddr`.
+    let mut dst: sockaddr_storage = unsafe { mem::zeroed() };
+    match remote {
+        IpAddr::V4(ip) => {
+            let sin = unsafe { &mut *ptr::from_mut(&mut dst).cast::<sockaddr_in>() };
+            sin.sin_len = size_of::<sockaddr_in>()
+                .try_into()
+                .map_err(|_| default_err())?;
+            sin.sin_family = AF_INET.try_into().map_err(|_| default_err())?;
+            sin.sin_addr.s_addr = u32::from_ne_bytes(ip.octets());
+        }
+        IpAddr::V6(ip) => {
+            let sin6 = unsafe { &mut *ptr::from_mut(&mut dst).cast::<sockaddr_in6>() };
+            sin6.sin6_len = size_of::<sockaddr_in6>()
+                .try_into()
+                .map_err(|_| default_err())?;
+            sin6.sin6_family = AF_INET6.try_into().map_err(|_| default_err())?;
+            sin6.sin6_addr.s6_addr = ip.octets();
+        }
+    };
+
+    // Prepare route message structure.
+    let mut rtm: rt_msghdr = unsafe { mem::zeroed() };
+    rtm.rtm_msglen = (size_of::<rt_msghdr>() + dst.ss_len as usize)
+        .try_into()
+        .map_err(|_| default_err())?; // Length includes sockaddr
+    rtm.rtm_version = RTM_VERSION.try_into().map_err(|_| default_err())?;
+    rtm.rtm_type = RTM_GET.try_into().map_err(|_| default_err())?;
+    rtm.rtm_seq = fd.as_raw_fd(); // Abuse file descriptor as sequence number, since it's unique
+    rtm.rtm_addrs = RTA_DST | RTA_IFP; // Query for destination and obtain interface info
+
+    // Copy route message and destination `sockaddr` into message buffer.
+    let mut msg: Vec<u8> = vec![0; rtm.rtm_msglen as usize];
+    unsafe {
+        ptr::copy_nonoverlapping(
+            ptr::from_ref(&rtm).cast(),
+            msg.as_mut_ptr(),
+            size_of::<rt_msghdr>(),
+        );
+        ptr::copy_nonoverlapping(
+            ptr::from_ref(&dst).cast(),
+            msg.as_mut_ptr().add(size_of::<rt_msghdr>()),
+            dst.ss_len as usize,
+        );
+    }
+
+    // Send route message.
+    let res = unsafe { write(fd.as_raw_fd(), msg.as_ptr().cast(), msg.len()) };
```

Comment:
Do I understand correctly that `libc::write` is never partial? In other words, can one assume if `res` is not `-1`, the full `msg` was written?